lib/secrets/src/Cardano/Wallet/Primitive/Passphrase/Legacy.hs

```diff
@@ -63,34 +63,89 @@ import qualified Codec.CBOR.Write as CBOR
 import qualified Data.ByteString.Char8 as B8
-import Data.Either.Extra
-    ( eitherToMaybe
-    )
 
 #if HAVE_SCRYPT
 import Crypto.Scrypt
     ( EncryptedPass (..)
     , Pass (..)
+    , Salt (Salt)
+    , encryptPass
+    , scryptParamsLen
     , verifyPass'
     )
 
--- | Verify a wallet spending password using the legacy Byron scrypt encryption
--- scheme.
-checkPassphrase :: Passphrase "encryption" -> PassphraseHash -> Maybe Bool
-checkPassphrase pwd stored = Just $
-    verifyPass' (Pass (BA.convert (cborify pwd))) encryptedPass
-  where
-    encryptedPass = EncryptedPass (BA.convert stored)
-
 haveScrypt :: Bool
 haveScrypt = True
+
 #else
--- | Stub function for when compiled without @scrypt@.
-checkPassphrase :: Passphrase "encryption" -> PassphraseHash -> Maybe Bool
-checkPassphrase _ _ = Nothing
 
 haveScrypt :: Bool
 haveScrypt = False
+
+#endif
+
+{- NOTE [LegacyScrypt]
+
+We need to transition away from the unmaintained `scrypt` package.
+Specifically, the `scrypt` package is not supported on `aarch64-darwin`.
+
+We can use the `cryptonite` package instead.
+In order to ensure that the new package produces the same result as the
+old one, we proceed as follows:
+
+    * in production:
+        * use `scrypt` package if available
+        * use `cryptonite` on `aarch64-darwin`
+    * in testing:
+        * generate random passphrases,
+            encrypted with `scrypt` package if available.
+        * check that these encrypted passphrases
+            are verified correctly with the `cryptonite` package
+
+These tests ensure that the code using `cryptonite` can verify
+hashed passphrases that were created with `scrypt`.
+
+-}
+
+{-----------------------------------------------------------------------------
+    Passphrase hashing and verification using the Scrypt algorithm
+------------------------------------------------------------------------------}
+-- | Verify a wallet spending password using the legacy Byron scrypt encryption
+-- scheme.
+checkPassphrase :: Passphrase "encryption" -> PassphraseHash -> Bool
+checkPassphrase = checkPassphraseCryptonite
+
+-- | Encrypt a wallet spending password using
+-- the legacy Byron scrypt encryption scheme.
+--
+-- Do not use this function in production, only in unit tests!
+encryptPassphraseTestingOnly
+    :: MonadRandom m
+    => PassphraseHashLength
+    -> Passphrase "encryption"
+    -> m PassphraseHash
+#if HAVE_SCRYPT
+encryptPassphraseTestingOnly len pwd = mkPassphraseHash <$> genSalt
+  where
+    mkPassphraseHash :: Passphrase "salt" -> PassphraseHash
+    mkPassphraseHash (Passphrase salt) =
+        PassphraseHash
+        . BA.convert
+        . getEncryptedPass
+        $ encryptPass
+            params
+            (Salt $ BA.convert salt)
+            (Pass . BA.convert . unPassphrase $ cborify pwd)
+
+    params = case scryptParamsLen logN r p (fromIntegral len) of
+        Just x -> x
+        Nothing -> error "scryptParamsLen: unreachable code path"
+    logN = 14
```

Comment:
what about extracting Scrypt parameters, logN, r, p, outside scrypt and cryptonite functions and reuse the set in both places?

Reply:
Fair enough. I have defined the parameters `logN, r, p :: Int` as top-level declarations. This ensures that the functions `encryptPassphraseTestingOnly` and `encryptPassphraseCryptoniteWithLength` use the same numerical values.